langroid/parsing/parse_json.py

```diff
@@ -73,6 +74,31 @@ def add_quotes(s: str) -> str:
         return s
 
 
+def parse_imperfect_json(json_string: str) -> Union[Dict[str, Any], List[Any]]:
+    if not json_string.strip():
+        raise ValueError("Empty string is not valid JSON")
+
+    # First, try parsing with ast.literal_eval
+    try:
+        result = ast.literal_eval(json_string)
+        if isinstance(result, (dict, list)):
+            return result
+    except (ValueError, SyntaxError):
+        pass
+
+    # If ast.literal_eval fails or returns non-dict/list, try json.loads
+    try:
+        str = add_quotes(json_string)
+        result = json.loads(str)
```

Comment:
@nilspalumbo see some discussions here about preferring `ast.literal_eval`:

https://community.openai.com/t/response-from-tools-call-function-arguments-is-not-consistently-valid-json/594040/2

https://community.openai.com/t/function-calls-and-streaming/263393

In any case, I changed it to use a new `parse_imperfect_json()`  fn I created in `parse_json.py`: it first tries  `literal_eval` and then `json.loads`

Reply:
@nilspalumbo please see if the new tests in `test_json.py` address the scenarios that you had in mind when you made this change.

Reply:
Yeah, this works! I hadn't known about this issue, looks like ideally we'd want a parser which tolerates mixed Python/JSON syntax but this is enough for now.